core_impl/src/test/java/com/google/instrumentation/trace/SpanImplTest.java

```diff
@@ -103,20 +165,159 @@ public void toSpanData_EndedSpan() {
     testClock.advanceTime(Duration.create(0, 7777));
+    span.addAnnotation(Annotation.fromDescription(ANNOTATION_DESCRIPTION));
+    span.addAnnotation(ANNOTATION_DESCRIPTION, attributes);
+    NetworkEvent networkEvent =
+        NetworkEvent.builder(NetworkEvent.Type.RECV, 1).setMessageSize(3).build();
+    span.addNetworkEvent(networkEvent);
+    Link link = Link.fromSpanContext(spanContext, Link.Type.CHILD);
+    span.addLink(link);
     span.end(EndSpanOptions.builder().setStatus(Status.CANCELLED).build());
     Mockito.verify(startEndHandler, Mockito.times(1)).onEnd(span);
     SpanData spanData = span.toSpanData();
     assertThat(spanData.getContext()).isEqualTo(spanContext);
     assertThat(spanData.getDisplayName()).isEqualTo(SPAN_NAME);
     assertThat(spanData.getParentSpanId()).isEqualTo(parentSpanId);
     assertThat(spanData.getHasRemoteParent()).isFalse();
+    assertThat(spanData.getAnnotations().getDroppedEventsCount()).isEqualTo(0);
+    assertThat(spanData.getAnnotations().getEvents().size()).isEqualTo(2);
+    assertThat(spanData.getAnnotations().getEvents().get(0).getTimestamp())
+        .isEqualTo(timestamp.addNanos(7777));
+    assertThat(spanData.getAnnotations().getEvents().get(0).getEvent())
+        .isEqualTo(Annotation.fromDescription(ANNOTATION_DESCRIPTION));
+    assertThat(spanData.getAnnotations().getEvents().get(1).getTimestamp())
+        .isEqualTo(timestamp.addNanos(7777));
+    assertThat(spanData.getAnnotations().getEvents().get(1).getEvent())
+        .isEqualTo(Annotation.fromDescriptionAndAttributes(ANNOTATION_DESCRIPTION, attributes));
+    assertThat(spanData.getNetworkEvents().getDroppedEventsCount()).isEqualTo(0);
+    assertThat(spanData.getNetworkEvents().getEvents().size()).isEqualTo(1);
+    assertThat(spanData.getNetworkEvents().getEvents().get(0).getTimestamp())
+        .isEqualTo(timestamp.addNanos(7777));
+    assertThat(spanData.getNetworkEvents().getEvents().get(0).getEvent()).isEqualTo(networkEvent);
+    assertThat(spanData.getLinks().getDroppedLinksCount()).isEqualTo(0);
+    assertThat(spanData.getLinks().getLinks().size()).isEqualTo(1);
+    assertThat(spanData.getLinks().getLinks().get(0)).isEqualTo(link);
     assertThat(spanData.getStartTimestamp()).isEqualTo(timestamp);
     assertThat(spanData.getStatus()).isEqualTo(Status.CANCELLED);
     assertThat(spanData.getEndTimestamp()).isEqualTo(timestamp.addNanos(7777));
   }
+
+  @Test
+  public void droppingAnnotations() {
+    SpanImpl span =
+        SpanImpl.startSpan(
+            spanContext,
+            recordSpanOptions,
+            SPAN_NAME,
+            parentSpanId,
+            false,
+            TraceParams.DEFAULT,
+            startEndHandler,
+            timestampConverter,
+            testClock);
+    Annotation annotation = Annotation.fromDescription(ANNOTATION_DESCRIPTION);
+    final int maxNumberOfAnnotations = TraceParams.DEFAULT.getMaxNumberOfAnnotations();
```

Comment:
I think this test would be clearer if it set maxNumberOfAnnotations on the TraceParams and tested with a known limit.